tests/unit/lms/views/basic_lti_launch_test.py

```diff
@@ -289,47 +290,96 @@ def pyramid_request(self, pyramid_request):
         return pyramid_request
 
 
+class TestConfigureModuleItem(ConfiguredLaunch):
+    def make_request(self, context, pyramid_request):
+        BasicLTILaunchViews(context, pyramid_request).configure_module_item()
+
+    def test_it_saves_the_assignments_document_url_to_the_db(
+        self, context, pyramid_request, ModuleItemConfiguration
+    ):
+        self.make_request(context, pyramid_request)
+
+        ModuleItemConfiguration.set_document_url.assert_called_once_with(
+            pyramid_request.db,
+            "TEST_TOOL_CONSUMER_INSTANCE_GUID",
+            "TEST_RESOURCE_LINK_ID",
+            "TEST_DOCUMENT_URL",
+        )
+
+    def test_it_configures_via_url(self, context, pyramid_request, via_url):
+        self.make_request(context, pyramid_request)
+
+        via_url.assert_called_once_with(pyramid_request, "TEST_DOCUMENT_URL")
+        assert context.js_config["urls"]["via_url"] == via_url.return_value
+
+    def test_it_configures_frontend_grading(
+        self, context, pyramid_request, frontend_app, via_url, ModuleItemConfiguration,
+    ):
+        self.make_request(context, pyramid_request)
+
+        frontend_app.configure_grading.assert_called_once_with(
+            pyramid_request, context.js_config
+        )
+
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request):
+        pyramid_request.parsed_params = {
+            "document_url": "TEST_DOCUMENT_URL",
+            "resource_link_id": "TEST_RESOURCE_LINK_ID",
+            "tool_consumer_instance_guid": "TEST_TOOL_CONSUMER_INSTANCE_GUID",
+        }
+
+        return pyramid_request
+
+
 class TestUnconfiguredBasicLTILaunch:
     def test_it_sets_the_right_javascript_config_settings(
-        self, BearerTokenSchema, bearer_token_schema, context, pyramid_request
+        self, bearer_token_schema, context, pyramid_request
     ):
-        pyramid_request.params[
-            "custom_canvas_api_domain"
-        ] = "TEST_CUSTOM_CANVAS_API_DOMAIN"
+        pyramid_request.params.update({"some_random_rubbish": "SOME_RANDOM_RUBBISH"})
+
         pyramid_request.registry.settings["google_client_id"] = "TEST_GOOGLE_CLIENT_ID"
         pyramid_request.registry.settings[
             "google_developer_key"
         ] = "TEST_GOOGLE_DEVELOPER_KEY"
 
         BasicLTILaunchViews(context, pyramid_request).unconfigured_basic_lti_launch()
 
-        BearerTokenSchema.assert_called_once_with(pyramid_request)
-        bearer_token_schema.authorization_param.assert_called_once_with(
-            pyramid_request.lti_user
-        )
```

Comment:
We don't care about the internal implementation of this. It output is correct, we don't care how it happened.

White box testing like this can increase the cost refactoring later.

Reply:
Righto, I'll take another look at this. I still suspect this is a mocking issue as we don't seem to have the ability to add input that can generate a reliable output.

Reply:
My mistake here was assuming that the test as written actually check outputs that would matter if you broke the code. We don't really check it because of heavy mocking.

To be clear I wasn't attempting to fix this problem by just removing tests, I assumed (incorrectly) that the assertions here were unnecessary because the final assertion about the output would actually test the class.

So I think what I've said is it's `unconfigued_basic_lti_launch()`'s job to put a valid JWT token in the output. Whether it does this by calling `BearerTokenSchema` or not is an implementation detail. We provide inputs, run the code and check the outputs.

This also has the benefit of being able to read the tests and actually see what the end-point is supposed to do. I don't really believe in the unit-tests as documentation, but it's a nice direction to head for.

Reply:
My thoughts on isolated testing (note: "isolated" not mocked, mocks are just one possible end to the means that is isolation) here: https://seanh.cc/post/when-to-use-mocks/#isolated-tests-vs-integrated-tests. Or at least those are the thoughts from 2017 that drove the approach to unit testing that we currently have. I wouldn't say I believe much in tests as documentation either, but I do believe in unit tests as drivers of better code design (at least if you pay attention to what your unit tests are telling you, which isn't easy), as a practical way to get high coverage, to help with refactoring and change management, etc.

But whether and how that applies to views in particular I think is open to discussion.

I think what you're talking about here is integrated tests. Not necessarily full end-to-end tests (or what we call "functional" tests in Hypothesis) but tests where at least the view is tested in integration with `BearerTokenSchema` (in this case) rather than tested strictly in isolation from all its first-party dependencies (all first-party dependencies mocked out).

The current pattern for `tests/unit/` tests in Hypothesis apps is that those are isolated unit tests not integrated tests.

We did talk about adding a `tests/integrated/` and views might be a good kind of thing to add such integrated tests for.

But we also said that only `tests/unit/` would count towards coverage so you would still require isolated unit tests for the views as well to get 100% coverage.

Unless of course we come up with a different idea...